state-chain/pallets/cf-elections/src/electoral_systems/block_witnesser.rs

```diff
@@ -172,123 +169,104 @@ impl<
-
-			for root in
-				witness_range.clone().step_by(Into::<u64>::into(Chain::WITNESS_PERIOD) as usize)
-			{
-				log::info!("New election for root: {:?}", root);
-				ElectoralAccess::new_election(
-					(),
-					(
-						Chain::block_witness_range(root).into(),
-						ElectionGenerator::generate_election_properties(root),
-					),
-					(),
-				)?;
-				last_block_election_emitted_for = root;
-				open_elections = open_elections.saturating_add(1);
-			}
-
-			// NB: We do not clear any of the unprocessed data here. This is because we need to
-			// prevent double dispatches. By keeping the state, if we have a reorg we can check
-			// against the state in the process_block_data hook to ensure we don't double
-			// dispatch.
-		} else {
-			// ==== No reorg case ====
+		let mut remaining_election_identifiers = election_identifiers.clone();
 
-			// We could have multiple elections going, for different block/ranges.
-			for election_identifier in election_identifiers {
-				let election_access = ElectoralAccess::election_mut(election_identifier);
-				if let Some(block_data) = election_access.check_consensus()?.has_consensus() {
-					let (root_block_number, _extra_properties) = election_access.properties()?;
+		let last_seen_root = match chain_progress {
+			ChainProgress::WaitingForFirstConsensus => return Ok(()),
+			ChainProgress::Reorg(reorg_range) => {
+				// Delete any elections that are ongoing for any blocks in the reorg range.
+				for (i, election_identifier) in election_identifiers.into_iter().enumerate() {
+					let election = ElectoralAccess::election_mut(election_identifier);
+					let properties = election.properties()?;
+					if properties.0.into_range_inclusive() == *reorg_range {
+						election.delete();
+						open_elections = open_elections.saturating_sub(1);
+						remaining_election_identifiers.remove(i);
+					}
+				}
 
-					election_access.delete();
+				// TODO: Wrap with safe mode, no new elections.
+				for root in
+					reorg_range.clone().step_by(Into::<u64>::into(Chain::WITNESS_PERIOD) as usize)
+				{
+					log::info!("New election for root: {:?}", root);
+					ElectoralAccess::new_election(
+						(),
+						(
+							Chain::block_witness_range(root).into(),
+							ElectionGenerator::generate_election_properties(root),
+						),
+						(),
+					)?;
+					last_block_election_emitted_for = root;
+					open_elections = open_elections.saturating_add(1);
+				}
 
-					open_elections = open_elections.saturating_sub(1);
+				// NB: We do not clear any of the unprocessed data here. This is because we need to
+				// prevent double dispatches. By keeping the state, if we have a reorg we can check
+				// against the state in the process_block_data hook to ensure we don't double
+				// dispatch.
+				*reorg_range.end()
+			},
+			ChainProgress::None(last_block_root_seen) => *last_block_root_seen,
+			ChainProgress::Continuous(witness_range) => *witness_range.start(),
```

Comment:
Are you sure you want `.start()` here? You assign this to last seen root, so shouldn't this also be `.end()`? Of course this element should be a "witness root".

Reply:
yeah should be end 👍